pkg/storage/ingest/reader.go

```diff
@@ -191,46 +192,104 @@ func (r *PartitionReader) processNextFetches(ctx context.Context, delayObserver
 	r.notifyLastConsumedOffset(fetches)
 }
 
-func (r *PartitionReader) processNextFetchesUntilMaxLagHonored(ctx context.Context, maxLag time.Duration) error {
-	level.Info(r.logger).Log("msg", "partition reader is starting to consume partition until max consumer lag is honored", "max_lag", maxLag)
+// processNextFetchesUntilTargetOrMaxLagHonored process records from Kafka until at least the maxLag is honored.
+// This function does a best-effort to get lag below targetLag, but it's not guaranteed that it will be
+// reached once this function successfully returns (only maxLag is guaranteed).
+func (r *PartitionReader) processNextFetchesUntilTargetOrMaxLagHonored(ctx context.Context, targetLag, maxLag time.Duration) error {
+	logger := log.With(r.logger, "target_lag", targetLag, "max_lag", maxLag)
+	level.Info(logger).Log("msg", "partition reader is starting to consume partition until target and max consumer lag is honored")
 
+	attempts := []func() (currLag time.Duration, _ error){
+		// First process fetches until at least the max lag is honored.
+		func() (time.Duration, error) {
+			return r.processNextFetchesUntilLagHonored(ctx, maxLag, logger)
+		},
+
+		// If the target lag hasn't been reached with the first attempt (which stops once at least the max lag
+		// is honored) then we try to reach the (lower) target lag within a fixed time (best-effort).
+		// The timeout is equal to the max lag. This is done because we expect at least a 1x replay speed
+		// from Kafka (which means at most it takes 1s to ingest 1s of data).
+		func() (time.Duration, error) {
+			timedCtx, cancel := context.WithTimeoutCause(ctx, maxLag, errWaitTargetLagDeadlineExceeded)
+			defer cancel()
+
+			return r.processNextFetchesUntilLagHonored(timedCtx, targetLag, logger)
+		},
+
+		// If the target lag hasn't been reached with the previous attempt that we'll move on. However,
+		// we still need to guarantee that in the meanwhile the lag didn't increase and max lag is still honored.
+		func() (time.Duration, error) {
+			return r.processNextFetchesUntilLagHonored(ctx, maxLag, logger)
+		},
+	}
+
+	var currLag time.Duration
+	for _, attempt := range attempts {
+		var err error
+
+		currLag, err = attempt()
+		if errors.Is(err, errWaitTargetLagDeadlineExceeded) {
+			continue
+		}
+		if err != nil {
+			return err
+		}
+		if currLag <= targetLag {
+			level.Info(logger).Log(
+				"msg", "partition reader consumed partition and current lag is lower than configured target consumer lag",
+				"last_consumed_offset", r.consumedOffsetWatcher.LastConsumedOffset(),
+				"current_lag", currLag,
+			)
+			return nil
+		}
+	}
+
+	level.Warn(logger).Log(
```

Comment:
this is better off without using  `loggerWithCurrentLag`, right? Even a negative lag is informative. Maybe leave a tiny comment

Reply:
If we reach this point, the `currLag` is > `targetLag`. We could also use `loggerWithCurrentLag`, I think it doesn't matter much. The real purpose of `loggerWithCurrentLag` was making it conditional (only if set) so let me rename the function to make it more clear.

Reply:
> rename the function to make it more clear

https://github.com/grafana/mimir/pull/8579/commits/391985544ca6202b25aa4cddd93b7b183bc07eb4